toolchain/docs/check/pattern_matching.md

```diff
@@ -0,0 +1,285 @@
+it does pass a reference to the storage that `F` should construct the return
+value in. So although we would colloquially say that `F` takes two parameters of
+type `type`, it has a single `Call` parameter of type `Core.String`.
+
+If Carbon supports general patterns in function parameter lists, that introduces
+additional ways that `Call` parameters can diverge from the colloquial meaning.
+For example:
+
+```carbon
+fn G(x: i32, var (y: i32, z: i32));
+fn H(x: i32, (y: i32, var z: i32));
+```
+
+A `var` pattern converts the scrutinee to a durable reference expression, and
+then performs further pattern matching on the object it refers to. As a result,
+`G` has two `Call` parameters: a value corresponding to `x`, and a reference to
+an object of type `(i32, i32)`, corresponding to both `y` and `z`. On the other
+hand, `H` has 3 `Call` parameters: values corresponding to `x` and `y`, and a
+reference corresponding to `z`.
+
+### Caller and callee matching
+
+The `Call` parameters define the API boundary between the caller and callee at
+the SemIR level. As a result, responsibility for matching the arguments against
+the parameter list is split between the caller and the callee. Continuing the
+example from above, given the call `G(0, (x, y))`, the caller is responsible for
+converting `0` to `i32`, and for initializing a new `(i32, i32)` object from
+`(x, y)`, but the callee is responsible for binding the name `x` to its first
+`Call` parameter, and for destructuring its second `Call` parameter and binding
+the names `y` and `z` to its elements.
+
+In SemIR we represent this situation with special `ParamPattern` insts, which
+mark the boundary: there is exactly one `ParamPattern` inst for each `Call`
+parameter, which matches the entire corresponding `Call` argument. The
+subpatterns of the `ParamPattern`s are matched on the callee side, and
+everything above them is matched on the caller side. There are multiple kinds of
+`ParamPattern` inst, which correspond to different ways of passing a parameter
+(such as by reference or by value).
+
+When performing callee-side pattern matching, we do not have an actual scrutinee
+expression. Instead, for each `ParamPattern` inst we generate a corresponding
+`Param` inst, which reads from the corresponding entry in the `Call` argument
+list, and we use that as the scrutinee of the `ParamPattern`. Every
+`ParamPattern` kind has a corresponding `Param` kind.
+
+### The return slot
+
+If a function has a declared return type, the function takes an additional
+`Call` parameter, which points to the storage that should be initialized with
+the return value. This `Call` parameter is represented as an `OutParamPattern`
+inst with a `ReturnSlotPattern` inst as a subpattern. The `ReturnSlotPattern`
+also represents the return type declaration itself, such as in
+`SemIR::FunctionFields`. The SemIR that matches these patterns consists of a
```

Comment:
Why do you prefix some things with a namespace (`SemIR::` here, but also `Check::Context` since this is check documentation) but not others?

Reply:
I'm not sure why I qualified this one; I agree it's not necessary. I do think it's helpful to qualify `Check::Context` to help avoid potential confusion with `Parse::Context` -- I know it's not truly ambiguous because this is check documentation, but it could still be a temporary stumbling block for some readers.